File: cpp/gestores/gestor_periodos.cpp
#include "gestor_periodos.h"
#include <algorithm>

namespace pita {

static std::string a_mayusculas(std::string s) {
    std::transform(s.begin(), s.end(), s.begin(), [](unsigned char c) { return std::toupper(c); });
    return s;
}

GestorPeriodosAcademicos::GestorPeriodosAcademicos(
    ListaEnlazada<PeriodoAcademico>& periodos,
    ListaEnlazada<OfertaCurso>& ofertas
) : periodos(periodos),
    ofertas(ofertas) {}

int GestorPeriodosAcademicos::siguienteId() {
    int maxId = 0;
    for (const auto& p : periodos) {
        if (p.idPeriodo.has_value() && *p.idPeriodo > maxId) {
            maxId = *p.idPeriodo;
        }
    }
    return maxId + 1;
}

PeriodoAcademico* GestorPeriodosAcademicos::buscar(int idPeriodo) {
    for (auto& p : periodos) {
        if (p.idPeriodo.has_value() && *p.idPeriodo == idPeriodo) {
            return &p;
        }
    }
    throw ErrorPeriodo("No existe el periodo con ID " + std::to_string(idPeriodo));
}

void GestorPeriodosAcademicos::validarFechas(const PeriodoAcademico& p) {
    if (p.fechaInicio.has_value() && p.fechaFin.has_value() && !p.fechaInicio->empty() && !p.fechaFin->empty()) {
        if (*p.fechaFin < *p.fechaInicio) {
            throw ErrorPeriodo("La fecha final no puede preceder a la fecha inicial");
        }
    }
    if (p.fechaInicioMatricula.has_value() && p.fechaFinMatricula.has_value() && !p.fechaInicioMatricula->empty() && !p.fechaFinMatricula->empty()) {
        if (*p.fechaFinMatricula < *p.fechaInicioMatricula) {
            throw ErrorPeriodo("La ventana de matricula tiene fechas invalidas");
        }
    }
    if (p.fechaLimiteCancelacion.has_value() && p.fechaFin.has_value() && !p.fechaLimiteCancelacion->empty() && !p.fechaFin->empty()) {
        if (*p.fechaLimiteCancelacion > *p.fechaFin) {
            throw ErrorPeriodo("La fecha limite de cancelacion no puede superar el fin del periodo");
        }
    }
}

PeriodoAcademico& GestorPeriodosAcademicos::crearPeriodo(PeriodoAcademico periodo) {
    validarFechas(periodo);
    if (!periodo.idPeriodo.has_value()) {
        periodo.idPeriodo = siguienteId();
    } else {
        for (const auto& item : periodos) {
            if (item.idPeriodo.has_value() && *item.idPeriodo == *periodo.idPeriodo) {
                throw ErrorPeriodo("Ya existe el periodo " + std::to_string(*periodo.idPeriodo));
            }
        }
    }

    if (!periodo.estado.has_value() || periodo.estado->empty()) {
        periodo.estado = "ABIERTO";
    }

    periodos.push_back(std::move(periodo));
    return periodos.back();
}

PeriodoAcademico& GestorPeriodosAcademicos::cerrarPeriodo(int idPeriodo) {
    PeriodoAcademico* p = buscar(idPeriodo);
    p->estado = "CERRADO";
    return *p;
}

PeriodoAcademico& GestorPeriodosAcademicos::abrirPeriodo(int idPeriodo) {
    PeriodoAcademico* p = buscar(idPeriodo);
    if (p->estado.has_value() && a_mayusculas(*p->estado) == "CERRADO") {
        throw ErrorPeriodo("No se puede abrir un periodo cerrado");
    }
    p->estado = "ABIERTO";
    return *p;
}

PeriodoAcademico* GestorPeriodosAcademicos::consultarPeriodo(int idPeriodo) {
    return buscar(idPeriodo);
}

ListaEnlazada<PeriodoAcademico> GestorPeriodosAcademicos::listarPeriodos(bool incluirCerrados) const {
    if (incluirCerrados) return periodos;
    ListaEnlazada<PeriodoAcademico> res;
    for (const auto& p : periodos) {
        if (!p.estado.has_value() || a_mayusculas(*p.estado) != "CERRADO") {
            res.push_back(p);
        }
    }
    return res;
}

ListaEnlazada<PeriodoAcademico> GestorPeriodosAcademicos::consultarPeriodos(const std::string& estado) const {
    if (estado.empty()) return periodos;
    std::string estBuscado = a_mayusculas(estado);
    ListaEnlazada<PeriodoAcademico> res;
    for (const auto& p : periodos) {
        if (p.estado.has_value() && a_mayusculas(*p.estado) == estBuscado) {
            res.push_back(p);
        }
    }
    return res;
}

PeriodoAcademico& GestorPeriodosAcademicos::modificarPeriodo(int idPeriodo, const PeriodoAcademico& datosActualizados) {
    PeriodoAcademico* actual = buscar(idPeriodo);
    validarFechas(datosActualizados);
    PeriodoAcademico copia = datosActualizados;
    copia.idPeriodo = actual->idPeriodo;
    *actual = copia;
    return *actual;
}

ListaEnlazada<OfertaCurso> GestorPeriodosAcademicos::consultarOfertasPeriodo(int idPeriodo) {
    buscar(idPeriodo);
    ListaEnlazada<OfertaCurso> res;
    for (const auto& o : ofertas) {
        if (o.idPeriodo.has_value() && *o.idPeriodo == idPeriodo) {
            res.push_back(o);
        }
    }
    return res;
}

OfertaCurso& GestorPeriodosAcademicos::modificarOferta(int idOferta, int idPeriodo, const OfertaCurso& cambios) {
    OfertaCurso* oferta = nullptr;
    for (auto& o : ofertas) {
        if (o.idOfertaCurso.has_value() && *o.idOfertaCurso == idOferta) {
            oferta = &o;
            break;
        }
    }
    if (!oferta) {
        throw ErrorPeriodo("No existe la oferta con ID " + std::to_string(idOferta));
    }
    if (!oferta->idPeriodo.has_value() || *oferta->idPeriodo != idPeriodo) {
        throw ErrorPeriodo("La oferta no pertenece a ese periodo");
    }

    if (cambios.cupoMaximo.has_value() && *cambios.cupoMaximo < 0) {
        throw ErrorPeriodo("El cupo maximo no puede ser negativo");
    }
    if (cambios.cupoDisponible.has_value() && *cambios.cupoDisponible < 0) {
        throw ErrorPeriodo("El cupo disponible no puede ser negativo");
    }
    if (cambios.cupoMaximo.has_value() && cambios.cupoDisponible.has_value() &&
        *cambios.cupoDisponible > *cambios.cupoMaximo) {
        throw ErrorPeriodo("El cupo disponible no puede superar el cupo maximo");
    }

    // Actualizar campos preservando IDs
    if (cambios.grupo.has_value()) oferta->grupo = cambios.grupo;
    if (cambios.cupoMaximo.has_value()) oferta->cupoMaximo = cambios.cupoMaximo;
    if (cambios.cupoDisponible.has_value()) oferta->cupoDisponible = cambios.cupoDisponible;
    if (cambios.modalidad.has_value()) oferta->modalidad = cambios.modalidad;
    if (cambios.aula.has_value()) oferta->aula = cambios.aula;
    if (cambios.sede.has_value()) oferta->sede = cambios.sede;
    if (cambios.fechaInicio.has_value()) oferta->fechaInicio = cambios.fechaInicio;
    if (cambios.fechaFin.has_value()) oferta->fechaFin = cambios.fechaFin;
    if (cambios.estado.has_value()) oferta->estado = cambios.estado;

    return *oferta;
}

} // namespace pita

File: gestores/gestor_periodos.py
from __future__ import annotations

from datetime import date
from dataclasses import fields
from typing import Any

from dominio.modelo_datos import PeriodoAcademico


class ErrorPeriodo(ValueError):
    """Error de un periodo académico."""


class GestorPeriodosAcademicos:
    def __init__(self, periodos: list[PeriodoAcademico], ofertas: list[Any] | None = None) -> None:
        self.periodos = periodos
        self._ofertas = ofertas if ofertas is not None else []

    def crear_periodo(self, periodo: PeriodoAcademico) -> PeriodoAcademico:
        self._validar_fechas(periodo)
        periodo.idPeriodo = periodo.idPeriodo or self._siguiente_id()
        if any(item.idPeriodo == periodo.idPeriodo for item in self.periodos):
            raise ErrorPeriodo(f"Ya existe el periodo {periodo.idPeriodo}")
        if periodo.estado is None:
            periodo.estado = "ABIERTO"
        self.periodos.append(periodo)
        return periodo

    def cerrar_periodo(self, id_periodo: int) -> PeriodoAcademico:
        periodo = self._buscar(id_periodo)
        periodo.estado = "CERRADO"
        return periodo

    def consultar_periodo(self, id_periodo: int) -> PeriodoAcademico:
        return self._buscar(id_periodo)

    def listar_periodos(self, *, incluir_cerrados: bool = True) -> list[PeriodoAcademico]:
        if incluir_cerrados:
            return list(self.periodos)
        return [item for item in self.periodos if str(item.estado).upper() != "CERRADO"]

    def modificar_periodo(self, id_periodo: int, **cambios: Any) -> PeriodoAcademico:
        periodo = self._buscar(id_periodo)
        campos = set(getattr(periodo, "__dataclass_fields__", {}))
        desconocidos = set(cambios) - campos
        if desconocidos:
            raise ErrorPeriodo(f"Campos no válidos: {', '.join(sorted(desconocidos))}")
        anteriores = {campo: getattr(periodo, campo) for campo in cambios}
        for campo, valor in cambios.items():
            setattr(periodo, campo, valor)
        try:
            self._validar_fechas(periodo)
        except ErrorPeriodo:
            for campo, valor in anteriores.items():
                setattr(periodo, campo, valor)
            raise
        return periodo

    @staticmethod
    def _validar_fechas(periodo: PeriodoAcademico) -> None:
        if periodo.fechaInicio and periodo.fechaFin and periodo.fechaFin < periodo.fechaInicio:
            raise ErrorPeriodo("La fecha final no puede preceder a la fecha inicial")
        if periodo.fechaInicioMatricula and periodo.fechaFinMatricula and periodo.fechaFinMatricula < periodo.fechaInicioMatricula:
            raise ErrorPeriodo("La ventana de matrícula tiene fechas inválidas")
        if periodo.fechaLimiteCancelacion and periodo.fechaFin and periodo.fechaLimiteCancelacion > periodo.fechaFin:
            raise ErrorPeriodo("La fecha límite de cancelación no puede superar el fin del periodo")

    def _buscar(self, id_periodo: int) -> PeriodoAcademico:
        periodo = next((item for item in self.periodos if item.idPeriodo == id_periodo), None)
        if periodo is None:
            raise ErrorPeriodo(f"No existe el periodo con ID {id_periodo}")
        return periodo

    def consultar_periodos(self, *, estado: str | None = None) -> list[PeriodoAcademico]:
        """Consultar periodos con filtro opcional por estado."""
        if estado is None:
            return list(self.periodos)
        return [p for p in self.periodos if str(p.estado or "").upper() == estado.upper()]

    def consultar_ofertas_periodo(self, id_periodo: int) -> list[Any]:
        """Devuelve las ofertas completas de un período existente."""
        self._buscar(id_periodo)
        return list(self._ofertas_por_periodo(id_periodo))

    def modificar_oferta(self, id_oferta: int, id_periodo: int, **cambios: Any) -> Any:
        """Modifica una oferta del período sin permitir inconsistencias de cupo."""
        oferta = next((o for o in self._ofertas if o.idOfertaCurso == id_oferta), None)
        if oferta is None:
            raise ErrorPeriodo(f"No existe la oferta con ID {id_oferta}")
        if oferta.idPeriodo != id_periodo:
            raise ErrorPeriodo("La oferta no pertenece a ese periodo")
        permitidos = {campo.name for campo in fields(oferta)} - {"idOfertaCurso", "idPeriodo", "idCurso"}
        desconocidos = set(cambios) - permitidos
        if desconocidos:
            raise ErrorPeriodo(f"Campos de oferta no válidos: {', '.join(sorted(desconocidos))}")
        anteriores = {campo: getattr(oferta, campo) for campo in cambios}
        for campo, valor in cambios.items():
            setattr(oferta, campo, valor)
        try:
            if oferta.cupoMaximo is not None and oferta.cupoMaximo < 0:
                raise ErrorPeriodo("El cupo máximo no puede ser negativo")
            if oferta.cupoDisponible is not None and oferta.cupoDisponible < 0:
                raise ErrorPeriodo("El cupo disponible no puede ser negativo")
            if (oferta.cupoMaximo is not None and oferta.cupoDisponible is not None
                    and oferta.cupoDisponible > oferta.cupoMaximo):
                raise ErrorPeriodo("El cupo disponible no puede superar el cupo máximo")
        except ErrorPeriodo:
            for campo, valor in anteriores.items():
                setattr(oferta, campo, valor)
            raise
        return oferta

    def abrir_periodo(self, id_periodo: int) -> PeriodoAcademico:
        """Abrir un periodo académico para permitir operaciones."""
        periodo = self._buscar(id_periodo)
        if str(periodo.estado or "").upper() == "CERRADO":
            raise ErrorPeriodo("No se puede abrir un periodo cerrado")
        periodo.estado = "ABIERTO"
        return periodo

    def _ofertas_por_periodo(self, id_periodo: int) -> list:
        if self._ofertas is None:
            return []
        return [o for o in self._ofertas if o.idPeriodo == id_periodo]

    def _siguiente_id(self) -> int:
        return max((item.idPeriodo or 0 for item in self.periodos), default=0) + 1
